picky/src/jose/jwe.rs

```diff
@@ -455,58 +622,78 @@ enum EncoderMode<'a> {
     Direct(&'a [u8]),
 }
 
-fn encode_impl(jwe: Jwe, mode: EncoderMode) -> Result<String, JweError> {
-    let mut header = jwe.header;
-    let protected_header_base64 = general_purpose::URL_SAFE_NO_PAD.encode(serde_json::to_vec(&header)?);
+fn encode_impl(mut jwe: Jwe, mode: EncoderMode) -> Result<String, JweError> {
+    use picky_asn1_x509::PublicKey as RfcPublicKey;
 
     let (encrypted_key_base64, jwe_cek) = match mode {
         EncoderMode::Direct(symmetric_key) => {
-            if symmetric_key.len() != header.enc.key_size() {
+            if symmetric_key.len() != jwe.header.enc.key_size() {
                 return Err(JweError::InvalidSize {
                     ty: "symmetric key",
-                    expected: header.enc.key_size(),
+                    expected: jwe.header.enc.key_size(),
                     got: symmetric_key.len(),
                 });
             }
 
             // Override `alg` header with "dir"
-            header.alg = JweAlg::Direct;
```

Comment:
I found and fixed a bug here -- we were setting JweAlg::Direct for direct encryption, but that value was never serialized as the header was serializing to base64 in the code above.